src/server/game/World/WorldState.cpp

```diff
@@ -0,0 +1,289 @@
+ * Free Software Foundation; either version 3 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "MapMgr.h"
+#include "Player.h"
+#include "SharedDefines.h"
+#include "WorldState.h"
+
+WorldState* WorldState::instance()
+{
+    static WorldState instance;
+    return &instance;
+}
+
+WorldState::WorldState() : _isMagtheridonHeadSpawnedHorde(false), _isMagtheridonHeadSpawnedAlliance(false)
+{
+    _transportStates[CONDITION_THE_PURPLE_PRINCESS] = ZEPPELIN_STATE_UNKOWN;
+    _transportStates[CONDITION_THE_IRON_EAGLE]      = ZEPPELIN_STATE_UNKOWN;
+    _transportStates[CONDITION_THE_THUNDERCALLER]   = ZEPPELIN_STATE_UNKOWN;
+}
+
+WorldState::~WorldState()
+{
+}
+
+bool WorldState::IsConditionFulfilled(uint32 conditionId, uint32 state) const
+{
+    switch (conditionId)
+    {
+        case CONDITION_TROLLBANES_COMMAND:
+            return _isMagtheridonHeadSpawnedAlliance;
+        case CONDITION_NAZGRELS_FAVOR:
+            return _isMagtheridonHeadSpawnedHorde;
+        default:
+            return _transportStates.at(conditionId) == state;
+    }
+}
+
+void WorldState::HandleConditionStateChange(uint32 conditionId, uint32 state)
+{
+    _transportStates[conditionId] = state;
+}
+
+void WorldState::HandleExternalEvent(uint32 eventId, uint32 param)
+{
+    std::lock_guard<std::mutex> guard(_mutex);
+    switch (eventId)
+    {
+        case CUSTOM_EVENT_ADALS_SONG_OF_BATTLE:
+            if (!_adalSongOfBattleTimer)
+            {
+                _adalSongOfBattleTimer = 120 * MINUTE * IN_MILLISECONDS;
+                BuffAdalsSongOfBattle();
+            }
+            break;
+        case CUSTOM_EVENT_MAGTHERIDON_HEAD_SPAWN:
+            if (param == TEAM_ALLIANCE)
+            {
+                _isMagtheridonHeadSpawnedAlliance = true;
+                BuffMagtheridonTeam(TEAM_ALLIANCE);
+            }
+            else
+            {
+                _isMagtheridonHeadSpawnedHorde = true;
+                BuffMagtheridonTeam(TEAM_HORDE);
+            }
+            break;
+        case CUSTOM_EVENT_MAGTHERIDON_HEAD_DESPAWN:
+            if (param == TEAM_ALLIANCE)
+            {
+                _isMagtheridonHeadSpawnedAlliance = false;
+                DispelMagtheridonTeam(TEAM_ALLIANCE);
+            }
+            else
+            {
+                _isMagtheridonHeadSpawnedHorde = false;
+                DispelMagtheridonTeam(TEAM_HORDE);
+            }
+            break;
+        default:
+            break;
+    }
+}
+
+void WorldState::Update(uint32 diff)
+{
+    std::lock_guard<std::mutex> guard(_mutex);
```

Comment:
I don't think it's required here. We are calling it from world update loop and there shouldn't be any parallel interactions with WorldState at this point.